### PR TITLE
Track local users from deleted rooms and send synthentic leaves after room purge

### LINKED ISSUE #18599: Room deletion (and purging) does not always send a leave down sync to clients
Related to https://github.com/element-hq/synapse/issues/18464

### Problem

Room deletion (v2, and possibly v1) will purge the room upon deletion. It will first send a bunch of kick events, which are intended to go to clients to let them know to remove the room from their room list. The problem is however immediately after sending these the room is purged from the database so unless you are federated or happened to be syncing exactly as the leave was sent, you will miss the leave event. The leave events are purged along with the room itself.

Subsequent syncs made after the room is purged never include the leave event, so any clients that did not sync will be stuck with broken state. 

### Proposed solution

Synapse should generate synthetic leave events for users with devices that existed at the point in time the room was alive. Rather than relying on events in the room which may be purged, the homeserver should keep a mapping of user, device and room_id of deleted rooms that can be popped as the device syncs to ensure they see a leave down the sync in all cases. This minimises the retention of data onto the server with the bare minimum required to execute the leaves, without leaving any metadata.

To keep in mind that quite like device messages, these rows could hang around forever if devices do not sync so a sensible cleanup period should be incorporated (perhaps a month, or so) where these rows are purged anyway.

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #18599 

This adds a new table to track local users who were part of a locally deleted room, to preserve the information needed to send a leave down sync so that the client can remove the room locally.

- [x] Database design
- [x] Storage of members
- [In Progress] Handle down legacy sync
- [ ] Handle down sliding sync
- [ ] Tests

### CODE REVIEW

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -1920,12 +1921,23 @@ async def shutdown_room(
...
+            user
+            for user in await self.store.get_local_users_related_to_room(room_id)
+            if user[1] in (Membership.JOIN, Membership.INVITE, Membership.KNOCK)
+        ]
 
+        # When deleting a room, we want to store the local membership state so that we
+        # can still send synthetic leaves down sync after the room has been purged (if indeed it has).
+        # We must do this prior to kicking as otherwise the current_state_events
+        # table will be empty.
+        delete_stream_id = await self.store.store_deleted_room_members(room_id)
```
Comment (MadLittleMods) [inline synapse/handlers/room.py:1935]: Doing this in a separate transaction from the membership change could result in inconsistent data (like where we've inserted into `deleted_room_members` but haven't actually left the room).

Also we might as well be copying the `instance_name`/`stream_id` directly from the leave event itself.

File: synapse/handlers/sliding_sync/room_lists.py [hunk 2]
```diff
@@ -1235,6 +1235,10 @@ async def _get_newly_joined_and_left_rooms_fallback(
                 )
             )
 
+        deleted_rooms = await self.store.get_deleted_rooms_for_user(
+            user_id, to_token.room_key.stream
+        )
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/room_lists.py:1240]: We should take into account the `from_token` and `to_token` here.

Currently, this will grab every deleted room on every sync and we could do a lot less work in incremental syncs if we only grab what's changed.

File: synapse/storage/databases/main/roommember.py [hunk 3]
```diff
@@ -1845,6 +1846,146 @@ def _get_room_participation_txn(
...
+        Returns a tuple of room_id, stream position.
+        """
+
+        def _get_deleted_rooms_for_user(
+            txn: LoggingTransaction,
+        ) -> list[(str, str, PersistedEventPosition)]:
+            sql = """
+                SELECT room_id, room_version, deleted_at_stream_id FROM deleted_room_members
+                WHERE user_id = ?
+                AND ? < deleted_at_stream_id
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/roommember.py:1911]: For sync, the `to_token`/`to_key` should be be inclusive.

The general rule is: 

```
> `from_token` and <= `to_token`
```

Additionally we should be filtering be `instance_name` in the token, see the following as example: https://github.com/element-hq/synapse/blob/481c4e2b5561cf1c4718ae6672e06d7a8fd154c0/synapse/storage/databases/main/state_deltas.py#L187-L231

File: synapse/handlers/sliding_sync/room_lists.py [hunk 4]
```diff
@@ -1235,6 +1235,10 @@ async def _get_newly_joined_and_left_rooms_fallback(
                 )
             )
 
+        deleted_rooms = await self.store.get_deleted_rooms_for_user(
+            user_id, to_token.room_key.stream
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/room_lists.py:1239]: This should be taking into account the whole `to_token.room_key` (`RoomStreamToken`) since it can hold stream positions of multiple instances, see https://github.com/element-hq/synapse/blob/481c4e2b5561cf1c4718ae6672e06d7a8fd154c0/synapse/types/__init__.py#L619-L639

This applies to all of the new database store functions.

File: synapse/storage/schema/main/delta/93/01_add_deleted_room_memebers.sql [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+-- License, or (at your option) any later version.
+--
+-- See the GNU Affero General Public License for more details:
+-- <https://www.gnu.org/licenses/agpl-3.0.html>.
+
+CREATE TABLE deleted_room_members (
+    room_id TEXT NOT NULL,
+    room_version TEXT NOT NULL,
+    user_id TEXT NOT NULL,
+    deleted_at_stream_id bigint NOT NULL
```
Comment (MadLittleMods) [inline synapse/storage/schema/main/delta/93/01_add_deleted_room_memebers.sql:18]: We should be storing the `instance_name` alongside the `stream_id` of the leave event so we can properly compare against the `RoomStreamToken`.

File: synapse/handlers/room.py (hunk 1 above)
Comment (MadLittleMods) [inline synapse/handlers/room.py:1935]: We definitely need some tests for this stuff (Sliding Sync, legacy sync, application service)

Including this edge case:

> We need to ensure to cover the edge case of a user who leaves a room before it is purged, but then tries to `/sync` after the purge with a `since` token that's before their leave. Right now, they wouldn't see the leave event as it's been purged.
> 
> So when purging a room, not only do you need to create a row in `deleted_rooms_users` for all the users that were joined to the room at the time, but all local users who *have ever been in the room*.
>
> *-- @anoadragon453, https://github.com/element-hq/synapse/issues/18599#issuecomment-3008568598*

File: synapse/handlers/sliding_sync/room_lists.py [hunk 6]
```diff
@@ -1295,6 +1299,32 @@ async def _get_newly_joined_and_left_rooms_fallback(
                     event_pos=last_membership_change_in_from_to_range.event_pos,
                     room_version_id=await self.store.get_room_version_id(room_id),
                 )
+        for room_id, room_version_id, stream_pos in deleted_rooms:
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/room_lists.py:1302]: We should have a nice explanation like the other sections in the function and add to the `1)`, `2)` overview at the top

File: synapse/handlers/sliding_sync/room_lists.py (hunk 2 above)
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/room_lists.py:1240]: Might as well move this lower to where it's used

File: synapse/storage/databases/main/roommember.py [hunk 7]
```diff
@@ -1845,6 +1846,146 @@ def _get_room_participation_txn(
...
+            """
+            txn.execute(sql, (user_id, stream_pos))
+            return [(r[0], r[1], PersistedEventPosition("master", r[2])) for r in txn]
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/roommember.py:1914]: Use constant `MAIN_PROCESS_INSTANCE_NAME`

```suggestion
            return [(r[0], r[1], PersistedEventPosition(MAIN_PROCESS_INSTANCE_NAME, r[2])) for r in txn]
```

File: synapse/storage/schema/main/delta/93/01_add_deleted_room_memebers.sql [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+--
+-- This program is free software: you can redistribute it and/or modify
+-- it under the terms of the GNU Affero General Public License as
+-- published by the Free Software Foundation, either version 3 of the
+-- License, or (at your option) any later version.
+--
+-- See the GNU Affero General Public License for more details:
+-- <https://www.gnu.org/licenses/agpl-3.0.html>.
+
+CREATE TABLE deleted_room_members (
```
Comment (MadLittleMods) [inline synapse/storage/schema/main/delta/93/01_add_deleted_room_memebers.sql:14]: Add a comment about what this holds, what its used for, and why its useful/necessary.

File: synapse/storage/databases/main/roommember.py [hunk 9]
```diff
@@ -1845,6 +1846,146 @@ def _get_room_participation_txn(
...
+        This should be run just before a room is deleted (before any
+        kicks are made).
+
+        Args:
+            room_id: the ID of the room
+        """
+
+        max = self.get_room_max_stream_ordering()
+
+        room_version = await self.get_room_version_id()
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/roommember.py:1865]: Why do we care about the `room_version` being stored in the `deleted_room_members` table?

Feels wrong to store this here and duplicates the source of truth (`rooms` table)

File: synapse/handlers/sync.py [hunk 10]
```diff
@@ -1829,7 +1829,7 @@ async def generate_sync_result(
             full_state,
         )
 
-        logger.debug(
+        logger.info(
```
Comment (MadLittleMods) [inline synapse/handlers/sync.py:1832]: Seems like this should go back to the debug log level

Also applies to other logs added/modified in this file.

File: synapse/handlers/sliding_sync/room_lists.py [hunk 11]
```diff
@@ -1235,6 +1235,10 @@ async def _get_newly_joined_and_left_rooms_fallback(
                 )
             )
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/room_lists.py:1237]: It looks like you've only updated the fallback path (`_get_newly_joined_and_left_rooms_fallback`)

We also need to take this logic into account with the new path.

File: synapse/handlers/sliding_sync/room_lists.py [hunk 12]
```diff
@@ -1295,6 +1299,32 @@ async def _get_newly_joined_and_left_rooms_fallback(
...
+                user_id,
+                room_id,
+            )
+            # Note we use RoomsForUserStateReset here as it's ideal for the purpose of a deleted
+            # room where the event that caused us to leave no longer exists.
+            newly_left_room_map[room_id] = RoomsForUserStateReset(
+                room_id=room_id,
+                sender=None,
+                membership=Membership.LEAVE,
+                event_id=None,
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/room_lists.py:1324]: This makes me question whether we should be storing the `event_id`/`sender` of the leave event.

We only really expect the `sender`/`event_id` to be `None` for state resets. I haven't gone down the whole rabbit hole to see the downstream effects of these fields. We should definitely explore this route first and come up with a reason why we can or can't do this. If we can't do this, we also need to update the doc strings and comments around the scenarios where `event_id`/`sender` can be `None`.

File: synapse/handlers/sync.py [hunk 13]
```diff
@@ -2484,7 +2490,7 @@ async def handle_room_entries(room_entry: "RoomSyncResultBuilder") -> None:
         return set(newly_joined_rooms), set(newly_left_rooms)
 
     async def _have_rooms_changed(
```
Comment (MadLittleMods) [inline synapse/handlers/sync.py:2492]: Note: I haven't evaluated the solution for legacy `/sync` in this round of review

File: synapse/storage/databases/main/appservice.py [hunk 14]
```diff
@@ -319,6 +319,75 @@ def _create_appservice_txn(txn: LoggingTransaction) -> AppServiceTransaction:
             "create_appservice_txn", _create_appservice_txn
         )
 
+    async def get_deleted_room_members_for_appservice(
+        self, room_stream_id: int
+    ) -> List[EventBase]:
+        # If we have a room_stream_id, let's see if there are any deleted events.
+        events: List[EventBase] = []
+        logger.info("Processing deleted rooms stream ID %s", room_stream_id)
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/appservice.py:327]: Remove debug log

File: synapse/handlers/appservice.py [hunk 15]
```diff
@@ -139,6 +139,24 @@ async def _notify_interested_services(self, max_token: RoomStreamToken) -> None:
 
                     events_by_room: Dict[str, List[EventBase]] = {}
                     for event in events:
+                        # We never want to send leave events from deleted rooms to the AS since the events
+                        # may have been purged by the time it's pushed to the AS. Instead, we handle this
+                        # elsewhere.
```
Comment (MadLittleMods) [inline synapse/handlers/appservice.py:144]: I haven't fully evaluated the application service side of things. From first glance, things seem a bit questionable but I'll wait for the next round of review.

Mostly just looked this PR from the Sliding Sync perspective.

**Comment by duucck** [comment]
This PR is useful, pls review this for merging.

**Comment by anoadragon453** [comment]
@duucck this PR is currently in the author, @Half-Shot's, court to update.

**Comment by Half-Shot** [comment]
Hey hey. Sorry this has taken so long. I had to park this due to other priorities coming up, so it's up for grabs. It's so far along now that I think it really just needs picking up to solve the last few outstanding bugs. However, at the moment this cannot be me.

Can the maintainers find someone who might be able to finish the last mile?

**Comment by Half-Shot** [comment]
Assigning @wrjlewis to find a new owner.
